core/trino-main/src/main/java/io/trino/sql/gen/columnar/BloomFilter.java

```diff
@@ -0,0 +1,192 @@
+                selectedPositionsCount += filter.test(block, position) ? 1 : 0;
+            }
+            return selectedPositionsCount;
+        }
+
+        @Override
+        public InputChannels getInputChannels()
+        {
+            return inputChannels;
+        }
+    }
+
+    public static final class SliceBloomFilter
+    {
+        private final long[] bloom;
+        private final int bloomSizeMask;
+
+        /**
+         * A Bloom filter for a set of Slice values.
+         * This is approx 2X faster than the Bloom filter implementations in ORC and parquet because
+         * it uses single hash function and uses that to set 3 bits within a 64 bit word.
+         * The memory footprint is up to (4 * values.size()) bytes, which is much smaller than maintaining a hash set of strings.
+         *
+         * @param values List of values used for filtering
+         */
+        public SliceBloomFilter(List<Slice> values)
+        {
+            int bloomSize = getBloomFilterSize(values.size());
+            bloom = new long[bloomSize];
+            bloomSizeMask = bloomSize - 1;
+            for (Slice value : values) {
+                long hashCode = XxHash64.hash(value);
+                // Set 3 bits in a 64 bit word
+                bloom[bloomIndex(hashCode)] |= bloomMask(hashCode);
+            }
+        }
+
+        private static int getBloomFilterSize(int valuesCount)
+        {
+            // Linear hash table size is the highest power of two less than or equal to number of values * 4. This means that the
+            // table is under half full, e.g. 127 elements gets 256 slots.
+            int hashTableSize = Integer.highestOneBit(valuesCount * 4);
+            // We will allocate 8 bits in the bloom filter for every slot in a comparable hash table.
+            // The bloomSize is a count of longs, hence / 8.
+            return Math.max(1, hashTableSize / 8);
+        }
+
+        public boolean test(VariableWidthBlock block, int position)
+        {
+            return contains(block.getRawSlice(), block.getRawSliceOffset(position), block.getSliceLength(position));
+        }
+
+        @VisibleForTesting
+        public boolean contains(Slice data)
+        {
+            return contains(data, 0, data.length());
+        }
+
+        private boolean contains(Slice data, int offset, int length)
+        {
+            long hashCode = XxHash64.hash(data, offset, length);
+            long mask = bloomMask(hashCode);
+            return mask == (bloom[bloomIndex(hashCode)] & mask);
+        }
+
+        private int bloomIndex(long hashCode)
+        {
+            // Lower 21 bits are not used by bloomMask
+            // These are enough for the maximum size array that will be used here
+            return (int) (hashCode & bloomSizeMask);
+        }
+
+        private static long bloomMask(long hashCode)
+        {
+            // returned mask sets 3 bits based on portions of given hash
```

Comment:
nit:

```
            // returned mask sets 3 bits based on portions of given hash
            // Extract 38th to 43rd bits
            return (1L << ((hashCode >> 21) & 63)) |
                    // Extract 32nd to 37th bits
                    (1L << ((hashCode >> 27) & 63)) |
                    // Extract 26th to 31st bits
                    (1L << ((hashCode >> 33) & 63));
```

Reply:
I think the convention in existing code has been to keep the operator at start of each new line
e.g.
```
TrinoFileStatus that = (TrinoFileStatus) o;
        return isDirectory == that.isDirectory
                && length == that.length
                && modificationTime == that.modificationTime
                && blockLocations.equals(that.blockLocations)
                && path.equals(that.path);
```